### PR TITLE
Additional codelists and new study_definition for project

### PR DESCRIPTION
**Pull request by NikNakk**
This is still a work in progress, since it can't be finalised until all of the codelists are in place. I'm therefore leaving this as a draft for now, and this is more a statement of intent rather than an indication that all is ready!

### CODE REVIEW

**Comment by NikNakk** [comment]
Some codelists are still not finalised. In particular, all of the high cost drug medication lists as well as mesalazine and the atopic dermatitis lists. The atopic dermatitis codes are still also CTV2 since the issue hasn't been updated yet

File: analysis/codelists.py
Comment (brianmackenna) [inline analysis/codelists.py]: The current version of protocol says we will only do a sensitivity analysis in SGSS

File: analysis/codelists.py
Comment (brianmackenna) [inline analysis/codelists.py]: As above re sgss

File: analysis/study_definition.py
Comment (brianmackenna) [inline analysis/study_definition.py:77]: I think the intention here is to produce 5 different columns here. There are a few things about his approach that we should reconsider
- we don't need all these for every single med
- therefore it doesn't meet principle for data minimisation
- creates extra data which can always add to run times
 - table 4 in protocol specified different dates for each one

File: analysis/study_definition.py
Comment (brianmackenna) [inline analysis/study_definition.py:43]: For meds in primary care this will be true but for the secondary care high costs medicines this will only be the last two years

File: analysis/study_definition.py
Comment (brianmackenna) [inline analysis/study_definition.py:105]: we generally don't include the date unless compelling reason for data minimisation principle

File: analysis/study_definition.py
Comment (brianmackenna) [inline analysis/study_definition.py:106]: do we need the date or just binary flag? Not clear from written protocol

File: analysis/study_definition.py
Comment (brianmackenna) [inline analysis/study_definition.py]: Protocol is for only ONS death - delete?

File: analysis/study_definition.py
Comment (brianmackenna) [inline analysis/study_definition.py]: we discuss using SGSS for a sensitivity analysis but not primary care. Reconsider?

File: analysis/study_definition.py
Comment (brianmackenna) [inline analysis/study_definition.py:4]: We should specify the included population in the python usings `patients.satisfying`. Something like

```
 population=patients.satisfying(
        """
    
        AND (age >=18 AND age <= 110)
        AND has_follow_up 
        AND (sex = "M" OR sex = "F")
        """,
```

File: analysis/study_definition.py
Comment (brianmackenna) [inline analysis/study_definition.py:10]: why do we need dates? Is binary yes/no enough? Checking from a data minimisation point of view and I can't see anything in protocol that requires it

File: analysis/study_definition.py
Comment (brianmackenna) [inline analysis/study_definition.py]: this is good idea BUT I'm pretty sure we aren't 100% consistent in the meds codelists and there might be further changes with a shift to new BNF coding system.

I should check with team but I think best thing might be for you to run and see hwere it breaks and we fix column names codelist by codelist  on OpenCodelists

File: analysis/study_definition.py
Comment (brianmackenna) [inline analysis/study_definition.py:249]: stp and msoa re not currently in the protocol.........

File: analysis/study_definition.py
Comment (brianmackenna) [inline analysis/study_definition.py]: not in protocol - delete?

File: codelists/codelists.txt
Comment (brianmackenna) [inline codelists/codelists.txt:14]: comment elsewhere but currently we just have sgss in protocol

**Review by brianmackenna — CHANGES_REQUESTED** [review]
@NikNakk this is a really phenomenal piece of work. Thank you! I've made some comments on the study definition. If you could address these, it will be in good shape for tech team review before merging to master. Just so you are aware your level of Python programming is more advanced than mine so perfectly acceptable to justify and keep.

My comments are generally classed in two areas

-  were we have deviated slightly from the written protocol (it might be that the protocol needs changing)
-  data minimisation - at OpenSAFELY we try to minimise data fields and requests where possible for privacy but also for efficiency reasons.

The [OpenSAFELY documentation](https://docs.opensafely.org/en/latest/) is much more up to date since you wrote this so it might be more helpful now.

I have not reviewed the high cost drugs section.............we should have a better idea on what the code will look like later this week after discussion between OpenSAFELY tech team and James.

File: analysis/codelists.py
Comment (NikNakk) [reply analysis/codelists.py]: Ok, we will remove

File: analysis/study_definition.py
Comment (NikNakk) [reply analysis/study_definition.py:105]: We do need the date for the STATA analysis, though it won't be exposed at the end of the STATA script to export

File: analysis/study_definition.py
Comment (NikNakk) [reply analysis/study_definition.py:106]: We do need the date for the STATA analysis, though it won't be exposed at the end of the STATA script to export

File: analysis/study_definition.py
Comment (NikNakk) [reply analysis/study_definition.py]: Ok, we will remove

File: analysis/study_definition.py
Comment (NikNakk) [reply analysis/study_definition.py:4]: Ok, will add.

File: analysis/study_definition.py
Comment (NikNakk) [reply analysis/study_definition.py:249]: We will add STP to the protocol and remove MSOA from the script.

File: analysis/study_definition.py
Comment (NikNakk) [reply analysis/study_definition.py]: IMD is in the protocol. We will remove rural_urban.

File: codelists/codelists.txt
Comment (NikNakk) [reply codelists/codelists.txt:14]: We will remove

File: analysis/study_definition.py
Comment (NikNakk) [reply analysis/study_definition.py]: Ok. The codelists are consistent for now since most of them were generated from the same source.

File: analysis/study_definition.py
Comment (NikNakk) [reply analysis/study_definition.py:10]: We compare dates of diagnosis to determine how to label people where diagnosis has changed.

File: analysis/study_definition.py
Comment (NikNakk) [reply analysis/study_definition.py]: Primary care removed.

File: analysis/study_definition.py
Comment (NikNakk) [reply analysis/study_definition.py:43]: We're no longer pulling the latest date since it's not used in the STATA, but I've left in the function in case it's helpful in future.

File: analysis/study_definition.py
Comment (NikNakk) [reply analysis/study_definition.py:77]: I've refined this to two columns for most meds and three for B cell therapies.

File: analysis/codelists.py
Comment (NikNakk) [reply analysis/codelists.py]: Removed

**Comment by NikNakk** [comment]
I've made the requested changes and/or commented against them. We're now also incorporating high cost drugs, but using a dummy file for now while the issue around invalid characters is sorted out.

**Comment by brianmackenna** [comment]
@evansd would you be able to give this code a review please?

I chatted with @sebbacon before Christmas and we think it would be good for someone from tech team to undertake a review before merging.

File: save_crossimd_codelists.sh
Comment (evansd) [inline save_crossimd_codelists.sh]: Is this script still used? It looks like the approach taken above is to keep non-opensafely codelists in a separate `crossimid-codelists` directory, which I think is the correct way of handling this.

The contents of the `codelists` dir needs to match the specification in `codelists.txt` otherwise the tests will fail.

**Review by evansd — APPROVED** [review]
This looks generally good. I don't have enough experience of Stata to sensibly review the Stata code, but the study def code all looks good (and I like the use of functions to remove repetition from the `StudyDefinition` itself).

Now the invalid char issue is fixed (see [#453](https://github.com/opensafely-core/cohort-extractor/pull/453)) it should be possible to include the full high cost drugs codelists, if they're ready.

We've got a new (simpler) test workflow now, so `.github/workflows/test_runner.yaml` should be updated to match [this](https://github.com/opensafely/research-template/blob/master/.github/workflows/test_runner.yaml) and `branch_release_deleter.yaml` is leftover legacy cruft and should be deleted. However, the tests won't run properly on this PR because its from outside of the OpenSAFELY org and so it won't have access to the STATA_LICENSE secret which it needs to run the Stata code. There's a workaround for this which involves someone in the OpenSAFELY org pushing the branch from their account and creating a "shadow PR" which will then cause the test status to update in the original PR.

File: save_crossimd_codelists.sh
Comment (NikNakk) [reply save_crossimd_codelists.sh]: No it's not needed any more. I've removed it with a new commit.

Comment (NikNakk) [reply save_crossimd_codelists.sh]: I've also now reverted the code change made to work around the invalid character issue above.

**Comment by brianmackenna** [comment]
I have merged the PR. I have amended write access for @myatesrheum and @NikNakk  to allow direct writing to the repo and run the tests directly.

**Pull request merged**
